File: 8_Octree/Octree.py
import open3d as o3d
import numpy as np
import matplotlib.pyplot as plt
import copy
import os
import sys
import time
import open3d_tutorial as o3dtut

print('Input')
N = 2000
pcd = o3dtut.get_armadillo_mesh().sample_points_poisson_disk(N)
# fit to unit cube
pcd.scale(1 / np.max(pcd.get_max_bound() - pcd.get_min_bound()),
          center=pcd.get_center())
pcd.colors = o3d.utility.Vector3dVector(np.random.uniform(0, 1, size=(N, 3)))
o3d.visualization.draw_geometries([pcd])

print('octree division')
octree = o3d.geometry.Octree(max_depth=4)
octree.convert_from_point_cloud(pcd, size_expand=0.01)
o3d.visualization.draw_geometries([octree], "Octree")

print('Voxelization')
voxel_grid = o3d.geometry.VoxelGrid.create_from_point_cloud(pcd, voxel_size=0.05)
o3d.visualization.draw_geometries([voxel_grid])

print('octree division')
octree = o3d.geometry.Octree(max_depth=4)
octree.create_from_voxel_grid(voxel_grid)
o3d.visualization.draw_geometries([octree])
#  an Octree can also be coverted to a VoxelGrid with to_voxel_grid.

def f_traverse(node, node_info):
    early_stop = False
    if isinstance(node, o3d.geometry.OctreeInternalNode):
        if isinstance(node, o3d.geometry.OctreeInternalPointNode):
            n = 0
            for child in node.children:
                if child is not None:
                    n += 1
            print(
                "{}{}: Internal node at depth {} has {} children and {} points ({})"
                .format('    ' * node_info.depth,
                        node_info.child_index, node_info.depth, n,
                        len(node.indices), node_info.origin))

            # we only want to process nodes / spatial regions with enough points
            early_stop = len(node.indices) < 250

    elif isinstance(node, o3d.geometry.OctreeLeafNode):
        if isinstance(node, o3d.geometry.OctreePointColorLeafNode):
            print("{}{}: Leaf node at depth {} has {} points with origin {}".
                  format('    ' * node_info.depth, node_info.child_index,
                         node_info.depth, len(node.indices), node_info.origin))
    else:
        raise NotImplementedError('Node type not recognized!')

    # early stopping: if True, traversal of children of the current node will be skipped
    return early_stop

octree = o3d.geometry.Octree(max_depth=4)
octree.convert_from_point_cloud(pcd, size_expand=0.01)
octree.traverse(f_traverse)

File: 8_Octree/src/Octree.cpp
#include <iostream>
#include <memory>
#include <thread>
#include <random>
#include <cstdlib>

#include <Eigen/Dense>

#include "open3d/Open3D.h"

int main(int argc, char* argv[]) 
{
    auto mesh = std::make_shared<open3d::geometry::TriangleMesh>();
    mesh = open3d::io::CreateMeshFromFile(argv[1]);  // Load Triangle Mesh
    //auto pcd = std::make_shared<open3d::geometry::PointCloud>();
    auto pcd = mesh -> SamplePointsPoissonDisk(2000);
    // Fit to unit cube
    Eigen::Vector3d ss = pcd->GetMaxBound() - pcd->GetMinBound();
    auto scale = ss.maxCoeff();
    std::cout << scale << std::endl;
    pcd ->Scale(1/scale, pcd->GetCenter());
    if (open3d::io::ReadTriangleMesh(argv[1], *mesh))
    {
        open3d::utility::LogInfo("Successfully read {}", argv[1]);
    }
    else
    {
        open3d::utility::LogInfo("Failed to read {}", argv[1]);
        return 1;
    }
    open3d::visualization::DrawGeometries({pcd});

    open3d::utility::LogInfo("Octree Division...");
    constexpr int max_depth = 4;
    auto octree = std::make_shared<open3d::geometry::Octree>(max_depth);
    octree-> ConvertFromPointCloud(*pcd);
    // open3d::visualization::DrawGeometries({octree});
    return 0;
} 